conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/processing/AccountsEventsProcessing.scala

```diff
@@ -0,0 +1,112 @@
+package tech.cryptonomic.conseil.indexer.tezos.processing
+
+import com.typesafe.scalalogging.LazyLogging
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+import scala.collection.immutable.SortedSet
+import tech.cryptonomic.conseil.common.config.ChainEvent
+import tech.cryptonomic.conseil.common.tezos.TezosTypes.BlockHash
+import tech.cryptonomic.conseil.indexer.tezos.{TezosIndexedDataOperations, TezosDatabaseOperations => TezosDb}
+import slick.jdbc.PostgresProfile.api._
+
+object AccountsEventsProcessing {
+
+  /** Events on which to trigger accounts updates, as ordered pairs of
+    * level and a pattern to decide which accounts needs to be reloaded.
+    */
+  type AccountUpdatesEvents = SortedSet[(Int, ChainEvent.AccountIdPattern)]
+
+}
+
+/** We capture global chain events that requires to refresh accounts
+  * data from the remote node.
+  * The original event to require such out-of-band processing was ad-hoc
+  * airdrop of mutez after a protocol upgrade.
+  * We identify when the refresh is needed as a specific block level, and
+  * restrict which accounts are affected via a pattern.
+  *
+  * @param db raw access to the underlying slick database
+  * @param indexedData access to the operations on locally indexed data
+  */
+class AccountsEventsProcessing(
+    db: Database,
+    indexedData: TezosIndexedDataOperations
+) extends LazyLogging {
+  import AccountsEventsProcessing._
+
+  /** Finds unprocessed levels for account refreshes (i.e. when there is a need to reload all accounts data from the chain) */
+  private[tezos] def unprocessedLevelsForRefreshingAccounts(
+      events: List[ChainEvent]
+  )(implicit ec: ExecutionContext): Future[AccountUpdatesEvents] =
+    events.collectFirst {
+      case ChainEvent.AccountsRefresh(levelsNeedingRefresh) if levelsNeedingRefresh.nonEmpty =>
+        db.run(TezosDb.fetchProcessedEventsLevels(ChainEvent.accountsRefresh.render)).map { levels =>
+          //used to remove processed events
+          val processed = levels.map(_.intValue).toSet
+          //we want individual event levels with the associated pattern, such that we can sort them by level
+          val unprocessedEvents = levelsNeedingRefresh.toList.flatMap {
+            case (accountPattern, levels) => levels.filterNot(processed).sorted.map(_ -> accountPattern)
+          }
+          SortedSet(unprocessedEvents: _*)
+        }
+    }.getOrElse(Future.successful(SortedSet.empty))
+
+  /* Possibly updates all accounts if the current block level is past any of the given ones
+   *
+   * @param events the relevant levels, each with its own selection pattern, that calls for a refresh
+   */
+  private[tezos] def processAccountRefreshes(events: AccountUpdatesEvents)(
```

Comment:
Hm... maybe it's worth to return different type, at the end, because when I first time checked this method, for me (from the signature perspective), it was doing nothing, but comment says something different.

Reply:
AccountUpdatesEvents is an alias to a `Set`.

Why would you expect that a method Set => Set would do nothing?
Even without considering side-effects... the signature being Set => F[Set]

Reply:
I explain that wrongly. What I had in mind is that, it's quiet easy to forget to fire this method up. When you have "chain" of types (so next method accepts type from previously returned) it's harder (or even impossible) to forget some method. Even during refactoring, moving code, copying it etc. When you have 2-3 methods with Set -> Set, does the order or running them matter? I hope not, but it easier to mess this up than, when you have proper types (not even aliases, because for the point mentioned above) it's even worst to me. I thought this one is a case class, but probably I should find out this one is an alias, by looking at the methods used on it.

By "would do nothing" I expect that whatever this method is doing for the same in&out types, there is hidden condition inside the method which sometimes does something with the data, sometimes does not. Like `list.filter`, `list.collect`, there are condition defined while executing these methods. When we are dealing with case classes, you also don't want to have `case class Account(name: Option[String] = None)` and inside of some methods do `account.copy(name = "account" + i)` cause that creates more confusion than helps, thus different type would be helpful here. Otherwise, we can create case class with optionals only and be happy about that, but as you know it's not the best pattern.

Reply:
I see now what you mean, specific types to address different states of a pipelined flow.
I'm not totally sold here as, foremost 'cause we're not deeply invested into types for reliability of the code. There will be gains in type-safety, but there's quite the cost in using types for everything.
I've found situations in this project where I could've done it, but I usually reject it because the additional complexity of the code needed wasn't worthwhile, all things considered.
I've usually made use of aliases and such, primarily to highlight the semantic of the logic, and to reduce code clutter, hence improving the reader experience and understanding. Obviously, I've found I have missed this goal not infrequently.

That said, I'll think about that part and try to see if something fits that would avoid the risks you outlined. I'm not promising I'll choose to do it, in the end.

Reply:
I'm mostly concerned by the fact that this set is passed recursively to the same method, hence I'm not so sure if I can really improve the situation.

Reply:
Ok I thing I finally gotcha here @ajuszczak

Reply:
Sure, what you are saying with complexity also make sense to me. There is always a tradeoff in such cases.